pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,222 @@
+}
+
+func (r *resourceNotReadyErr) Error() string {
+	return r.msg
+}
+
+// LocalObjectGetter is a function that is used to retrieve local object references
+type LocalObjectGetter func(ref corev1.LocalObjectReference, obj runtime.Object) error
+
+// AttributeReferencer is an interface for referencing and resolving
+// cross-resource attribute references. See
+// https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-cross-resource-referencing.md
+// for more information
+type AttributeReferencer interface {
+	// Build retrieves and builds the referenced attribute
+	Build(getter LocalObjectGetter) (value string, err error)
+
+	// Assign accepts a resource object, and assigns the given value to the corresponding property
+	Assign(mgd Managed, value string) error
+}
+
+// A ManagedReferenceResolver resolves the references to other managed resources
+type ManagedReferenceResolver interface {
+	ResolveReferences(context.Context, Managed) error
+}
+
+// ReferenceResolver resolves implements ManagedReferenceResolver interface
+type ReferenceResolver struct {
+	reader client.Reader
+}
+
+// NewReferenceResolver returns a new ReferenceResolver
+func NewReferenceResolver(c client.Reader) *ReferenceResolver {
+	return &ReferenceResolver{c}
+}
+
+// ResolveReferences resolves references made to other managed resources
+func (r *ReferenceResolver) ResolveReferences(ctx context.Context, mgd Managed) (err error) {
+	// this recovers from potential panics during execution of
+	// AttributeReferencer methods
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.Errorf("paniced while resolving references: %v", r)
+		}
+	}()
+
+	referencers, err := findRefResolverFields(mgd, false)
+	if err != nil {
+		return err
+	}
+
+	getter := r.getterFunc(ctx, mgd)
+	err = validateReferencesAreReady(getter, referencers)
+	if err != nil {
+		return err
+	}
+
+	for _, referencer := range referencers {
+		val, err := referencer.Build(getter)
+		if err != nil {
+			return err
+		}
+
+		err = referencer.Assign(mgd, val)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getterFunc returns a function to retrieve local references
+func (r *ReferenceResolver) getterFunc(ctx context.Context, mgd Managed) LocalObjectGetter {
+	return func(ref corev1.LocalObjectReference, obj runtime.Object) error {
+		err := r.reader.Get(ctx,
+			types.NamespacedName{
+				Namespace: mgd.GetNamespace(),
+				Name:      ref.Name,
+			},
+			obj)
+
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+// findRefResolverFields recursively finds all non-nil fields in a struct and its sub-structs
+// that implement AttributeReferencer and have `managed_resource:"reference_resolver"` tag
+func findRefResolverFields(obj interface{}, objHasTheRightTag bool) ([]AttributeReferencer, error) { // nolint:gocyclo
+
+	v := reflect.ValueOf(obj)
+	result := []AttributeReferencer{}
+
+	// ignore nil fields
```

Comment:
Nitpick: This is a case where I feel the comment explains something that is quite obvious from how the code reads. Even as someone unfamiliar with the `reflect` package this code reads like "is the kind of this thing a pointer, and is it nil? If so return early".